cpp/src/groupby/sort/group_single_pass_reduction_util.cuh

```diff
@@ -161,55 +183,119 @@ struct reduce_functor {
+      result->set_null_mask(std::move(null_mask), null_count);
+    }
+    return result;
+  }
+};
+
+template <aggregation::Kind K>
+struct group_reduction_functor<
+  K,
+  cudf::struct_view,
+  std::enable_if_t<is_group_redution_supported<K, cudf::struct_view>()>> {
+  static std::unique_ptr<column> invoke(column_view const& values,
+                                        size_type num_groups,
+                                        cudf::device_span<cudf::size_type const> group_labels,
+                                        rmm::cuda_stream_view stream,
+                                        rmm::mr::device_memory_resource* mr)
+  {
+    // This is be expected to be size_type.
+    using ResultType = cudf::detail::target_type_t<cudf::struct_view, K>;
+
+    auto result = make_fixed_width_column(
+      data_type{type_to_id<ResultType>()}, num_groups, mask_state::UNALLOCATED, stream, mr);
+
+    if (values.is_empty()) { return result; }
+
+    // When finding ARGMIN, we need to consider nulls as larger than non-null elements.
+    // Thing is opposite for ARGMAX.
+    auto const null_precedence =
+      (K == aggregation::ARGMIN) ? null_order::AFTER : null_order::BEFORE;
+    auto const flattened_values = structs::detail::flatten_nested_columns(
+      table_view{{values}}, {}, std::vector<null_order>{null_precedence});
+    auto const d_flattened_values_ptr = table_device_view::create(flattened_values, stream);
+    auto const flattened_null_precedences =
+      (K == aggregation::ARGMIN)
+        ? cudf::detail::make_device_uvector_async(flattened_values.null_orders(), stream)
+        : rmm::device_uvector<null_order>(0, stream);
+
+    // Perform segmented reduction to find ARGMIN/ARGMAX.
+    auto const do_reduction = [&](auto const& inp_iter, auto const& out_iter, auto const& binop) {
       thrust::reduce_by_key(rmm::exec_policy(stream),
                             group_labels.data(),
                             group_labels.data() + group_labels.size(),
-                            cudf::detail::make_validity_iterator(*valuesview),
+                            inp_iter,
                             thrust::make_discard_iterator(),
-                            validity.begin(),
+                            out_iter,
                             thrust::equal_to<size_type>{},
-                            thrust::logical_or<bool>{});
+                            binop);
+    };
+
+    auto const count_iter   = thrust::make_counting_iterator<ResultType>(0);
+    auto const result_begin = result->mutable_view().template begin<ResultType>();
+    if (values.has_nulls()) {
+      auto const binop = row_arg_minmax_fn<true>(values.size(),
+                                                 *d_flattened_values_ptr,
+                                                 flattened_null_precedences.data(),
+                                                 K == aggregation::ARGMIN);
+      do_reduction(count_iter, result_begin, binop);
```

Comment:
@davidwendt @jrhemstad I have new trouble and I'm asking for help!!!

Using `thrust::reduction_by_key` with a binop containing `row_lexicographic_comparator` significantly increases compile time and binary size. Compile time went from 1m to 8m on my local machine (size went from 2.8MB to 9.6MB) and stuck indefinitely in CI machines thus many CI tests failed because of timeout (https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/cudf/job/prb/job/cudf-cpu-cuda-build-arm64/CUDA=11.4/557/console and https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci-cuda115/job/cudf/job/prb/job/cudf-cpu-cuda-build/CUDA=11.5/272/console).

Since CI can't compile the new code, I have to rework it. I'm going to try removing the template `K` from this and make it a run time argument to see if it helps:
```
template <aggregation::Kind K>
struct group_reduction_functor
```

Do you guys have any other suggestions?  Thanks.

Reply:
I have no idea why CI machines failed to compile the new code. No error message from them:

cuda 11.4:
```
13:30:03 FAILED: CMakeFiles/cudf.dir/src/groupby/sort/group_argmin.cu.o 
```

and cuda 11.5:
```
13:54:56 Build timed out (after 40 minutes). Marking the build as failed.
```